### ISSUE TITLE
💅 noUnnecessaryConditions: false positive when a literal-initialised object property is mutated

### ISSUE THREAD

**Issue by XXMOHAMED012** [biomejs/biome#11174]
### Environment information

CLI:
  Version:                      2.5.6
  Color support:                false

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     1
  TERM:                         unset
  JS_RUNTIME_VERSION:           v24.16.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.3.14

Biome Configuration:
  Status:                       Loaded successfully.
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    suspicious/noUnnecessaryConditions

Workspace:
  Open Documents:               0

### Rule name

suspicious/noUnnecessaryConditions

### Playground link

https://biomejs.dev/playground/?lintRules=all&tab=diagnostics&code=YwBvAG4AcwB0ACAAZwB1AGEAcgBkACAAPQAgAHsAIABjAHUAcgByAGUAbgB0ADoAIABmAGEAbABzAGUAIAB9ADsACgAKAGUAeABwAG8AcgB0ACAAZgB1AG4AYwB0AGkAbwBuACAAcgB1AG4ATwBuAGMAZQAoACkAIAB7AAoACQBpAGYAIAAoAGcAdQBhAHIAZAAuAGMAdQByAHIAZQBuAHQAKQAgAHsACgAJAAkAcgBlAHQAdQByAG4AOwAKAAkAfQAKAAkAZwB1AGEAcgBkAC4AYwB1AHIAcgBlAG4AdAAgAD0AIAB0AHIAdQBlADsACgB9AAoACgAvAC8AIABDAG8AbgB0AHIAbwBsADoAIAB0AGgAZQAgAHMAYQBtAGUAIABsAG8AZwBpAGMAIABvAG4AIABhACAAYgBpAG4AZABpAG4AZwAgAHIAYQB0AGgAZQByACAAdABoAGEAbgAgAGEAIABwAHIAbwBwAGUAcgB0AHkAIABpAHMAIABuAG8AdAAgAGYAbABhAGcAZwBlAGQALgAKAGwAZQB0ACAAZgBsAGEAZwAgAD0AIABmAGEAbABzAGUAOwAKAAoAZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIAByAHUAbgBPAG4AYwBlAEIAaQBuAGQAaQBuAGcAKAApACAAewAKAAkAaQBmACAAKABmAGwAYQBnACkAIAB7AAoACQAJAHIAZQB0AHUAcgBuADsACgAJAH0ACgAJAGYAbABhAGcAIAA9ACAAdAByAHUAZQA7AAoAfQAKAA%3D%3D

### Expected result

No diagnostic should be emitted on line 4. `guard.current` is assigned `true` on line 7, so the condition is reachable.

### Reproduction

```ts
const guard = { current: false };

export function runOnce() {
    if (guard.current) {   // <-- "This condition is always falsy."
        return;
    }
    guard.current = true;
}

// Control: the same logic on a binding rather than a property is not flagged.
let flag = false;

export function runOnceBinding() {
    if (flag) {
        return;
    }
    flag = true;
}
```

`biome lint` on 2.5.6, with `suspicious/noUnnecessaryConditions` as the only enabled rule:

```
repro.ts:4:6 lint/suspicious/noUnnecessaryConditions ━━━━━━━━━━━━━━━━━━━━━━━━━━

  × This condition is always falsy.

    3 │ export function runOnce() {
  > 4 │     if (guard.current) {
      │         ^^^^^^^^^^^^^
    5 │         return;
    6 │     }

  i The value's type can never be truthy, so this check is redundant.

Found 1 error.
```

### Analysis

The property keeps the literal type `false` from its initialiser, and the later assignment is not taken into account. The control case in the same file shows the asymmetry: the mutation exemption added in #10108 covers plain bindings, but not member expressions. TypeScript widens the property to `boolean` here and reports nothing.

### Other shapes that reproduce

- `const r = useRef(false); … r.current = true` — this is the common React in-flight guard against double submission, and it is where I hit the false positive in real code.
- `const r = useRef<boolean>(false)` — an explicit type argument does not help; the initialiser still wins.
- `const r = useRef(0); … r.current = 1` — same behaviour, reported as always falsy.
- `const r = useRef(true); … r.current = false` — reported as always truthy.

An `as` cast on the initialiser (`useRef(false as boolean)`) silences it, which is consistent with the literal type being the cause.

### Impact

The reported code is not redundant, so acting on the diagnostic removes a working guard. There is no safe automatic fix; the only option today is a `biome-ignore` comment on every such condition.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
How does typescript-eslint behave towards this use case?

**Comment by XXMOHAMED012** [comment]
@typescript-eslint/no-unnecessary-condition does not report any of these.

Tested on typescript-eslint 8.65.0 / TypeScript 5.9.3, with the rule enabled on its own under type-aware linting:

const guard = { current: false };

export function runOnce() {
      if (guard.current) { return; }   // not reported
      guard.current = true;
}

export function useGuard() {
      const inFlight = useRef(false);
      return async function submit() {
              if (inFlight.current) { return; }   // not reported
              inFlight.current = true;
              // ...
      };
}

useRef(0) with n.current = 1 is likewise not reported. To confirm the rule was actually active in that run, the same file also contained if (s !== undefined) on a s: string parameter — that was reported ("Unnecessary conditional, the types have no overlap") and was the only diagnostic in the file.

The reason is the declared type, not mutation tracking. TypeScript widens the property of a mutable object literal to boolean, so the checker hands the rule boolean, never false. Confirmed with const x: false = guard.current, which is a type error. useRef<boolean>(false) produces the same widened type.

So typescript-eslint never reaches the situation described here — it does not need to reason about the later assignment, because the type it starts from was never the literal. That suggests the fix on the Biome side is at inference (widening a mutable property's literal initialiser) rather than extending the mutation exemption to member expressions.

**Comment by ematipico** [comment]
I would have appreciated just a playground link 😉

**Comment by XXMOHAMED012** [comment]
Fair 😄 — https://typescript-eslint.io/play/#fileType=.ts&code=PTAEGUBcFcCNQPYDNQCUCmBDAxpA5AM6jQHoZIA0oBCokAFgJZEB266AJq7QAKQCeAB3QFgAJyy5QjFnXrpQggDaZ%2BAczEJoLDgDoAUB3TYVE0Em25GCWSTLokAHgAqAPgAUMxpEaYlANT9odAAuUGcASjCAb1BsaDEJFkgw51AAXwBufX0QUABGXUUVGURYACtjSFAlb3QxPypBTWExAVAAW2hITEhOGt76-WwbAmq1aEwxDlAAXlBY%2BMT0ZLCkP1IM7P10AA9BBDbzSx8bUDFtAHkWbHR3CIX9UGkUdwmpvSWkyAfop%2BfzugYGIWNlnul-u9provitqvNIBd0NkIbkwAAmIoMBQSPwAWgA7oclDMMDhqgR6JhhGFMLIZHikLU1PRxpNpqBMGpMDIxqAOFpYEoFAQ4B1mARrCwdvtDtULDdTrZSABxdkce6PZ4jFh8mQAMWZrLmxFI5Hc6yUpAiYMBwNkmAI-Bux0VUuoYu8mr%2BAJeoE8LENjBZkBhCW%2Bv3%2Bvok9tt4Kj0kDRtDsOSJsRwTjdDE-C1vs5BJ51QACppxaRdBIaEoAG53G0J9LmGR%2BJS5n35g3JsPLNPzS2kLMQ8EonJ5ADMRRY0A69UY2EBSBlByOCqsZzsAGEtMlvf8dXzZPM7OaAAwN54xhKyNdK0CwGeCPe%2BxivFg9iN56NA69DhPv1N4QKW0sn0VE8m3ZJNCUMIGGYOIbA4bx3VxJQ22kIhtDYW4CAIKZ%2BCoOkZgASXAQEVz6PQ0VAYjqmaBA6yIbFzmgYUMOoAlvGweQZjOUo4KIZRVA0HcqL2CjXXXWQdURBAlHcAgwjGMQZDUSNnlff0iAAQlmY8dAcGROHUgErxBagQP%2BMzZAAIhs0cgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

no-unnecessary-condition reports only the control function at the bottom, which is there to show the rule is actually on. The object-property and useRef guards are clean.

**Comment by ematipico** [comment]
@XXMOHAMED012 can you please make sure the two playgrounds use the same content?

**Comment by XXMOHAMED012** [comment]
https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&tab=formatter&pane=Diagnostics&code=LwAvACAAUwB0AHUAYgAgAG8AZgAgAFIAZQBhAGMAdAAnAHMAIAB1AHMAZQBSAGUAZgAsACAAcwBvACAAdABoAGkAcwAgAG4AZQBlAGQAcwAgAG4AbwAgAEAAdAB5AHAAZQBzAC8AcgBlAGEAYwB0AC4ACgBkAGUAYwBsAGEAcgBlACAAZgB1AG4AYwB0AGkAbwBuACAAdQBzAGUAUgBlAGYAPABUAD4AKABpAG4AaQB0AGkAYQBsAFYAYQBsAHUAZQA6ACAAVAApADoAIAB7ACAAYwB1AHIAcgBlAG4AdAA6ACAAVAAgAH0AOwAKAAoALwAvACAAMQAuACAAcABsAGEAaQBuACAAbwBiAGoAZQBjAHQAIABsAGkAdABlAHIAYQBsACAAdwBoAG8AcwBlACAAcAByAG8AcABlAHIAdAB5ACAAaQBzACAAbQB1AHQAYQB0AGUAZAAgAGwAYQB0AGUAcgAKAGMAbwBuAHMAdAAgAGcAdQBhAHIAZAAgAD0AIAB7ACAAYwB1AHIAcgBlAG4AdAA6ACAAZgBhAGwAcwBlACAAfQA7AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAHIAdQBuAE8AbgBjAGUAKAApACAAewAKACAAIABpAGYAIAAoAGcAdQBhAHIAZAAuAGMAdQByAHIAZQBuAHQAKQAgAHsACgAgACAAIAAgAHIAZQB0AHUAcgBuADsACgAgACAAfQAKACAAIABnAHUAYQByAGQALgBjAHUAcgByAGUAbgB0ACAAPQAgAHQAcgB1AGUAOwAKAH0ACgAKAC8ALwAgADIALgAgAHQAaABlACAAcwBhAG0AZQAgAGcAdQBhAHIAZAAgAHYAaQBhACAAdQBzAGUAUgBlAGYAOgAgAHQAaABlACAAUgBlAGEAYwB0ACAAZABvAHUAYgBsAGUALQBzAHUAYgBtAGkAdAAgAHAAYQB0AHQAZQByAG4ACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAHUAcwBlAEcAdQBhAHIAZAAoACkAIAB7AAoAIAAgAGMAbwBuAHMAdAAgAGkAbgBGAGwAaQBnAGgAdAAgAD0AIAB1AHMAZQBSAGUAZgAoAGYAYQBsAHMAZQApADsACgAgACAAcgBlAHQAdQByAG4AIABhAHMAeQBuAGMAIABmAHUAbgBjAHQAaQBvAG4AIABzAHUAYgBtAGkAdAAoACkAIAB7AAoAIAAgACAAIABpAGYAIAAoAGkAbgBGAGwAaQBnAGgAdAAuAGMAdQByAHIAZQBuAHQAKQAgAHsACgAgACAAIAAgACAAIAByAGUAdAB1AHIAbgA7AAoAIAAgACAAIAB9AAoAIAAgACAAIABpAG4ARgBsAGkAZwBoAHQALgBjAHUAcgByAGUAbgB0ACAAPQAgAHQAcgB1AGUAOwAKACAAIAAgACAAdAByAHkAIAB7AAoAIAAgACAAIAAgACAAYQB3AGEAaQB0ACAAUAByAG8AbQBpAHMAZQAuAHIAZQBzAG8AbAB2AGUAKAApADsACgAgACAAIAAgAH0AIABmAGkAbgBhAGwAbAB5ACAAewAKACAAIAAgACAAIAAgAGkAbgBGAGwAaQBnAGgAdAAuAGMAdQByAHIAZQBuAHQAIAA9ACAAZgBhAGwAcwBlADsACgAgACAAIAAgAH0ACgAgACAAfQA7AAoAfQAKAAoALwAvACAAMwAuACAAbgB1AG0AZQByAGkAYwAgAHIAZQBmAAoAZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIAB1AHMAZQBDAG8AdQBuAHQAKAApACAAewAKACAAIABjAG8AbgBzAHQAIABuACAAPQAgAHUAcwBlAFIAZQBmACgAMAApADsACgAgACAAcgBlAHQAdQByAG4AIABmAHUAbgBjAHQAaQBvAG4AIABiAHUAbQBwACgAKQAgAHsACgAgACAAIAAgAGkAZgAgACgAbgAuAGMAdQByAHIAZQBuAHQAKQAgAHsACgAgACAAIAAgACAAIAByAGUAdAB1AHIAbgA7AAoAIAAgACAAIAB9AAoAIAAgACAAIABuAC4AYwB1AHIAcgBlAG4AdAAgAD0AIAAxADsACgAgACAAfQA7AAoAfQAKAAoALwAvACAAQwBvAG4AdAByAG8AbAAgAEEAOgAgAHQAaABlACAAcwBhAG0AZQAgAGwAbwBnAGkAYwAgAG8AbgAgAGEAIABiAGkAbgBkAGkAbgBnACAAcgBhAHQAaABlAHIAIAB0AGgAYQBuACAAYQAgAHAAcgBvAHAAZQByAHQAeQAuAAoAbABlAHQAIABmAGwAYQBnACAAPQAgAGYAYQBsAHMAZQA7AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAHIAdQBuAE8AbgBjAGUAQgBpAG4AZABpAG4AZwAoACkAIAB7AAoAIAAgAGkAZgAgACgAZgBsAGEAZwApACAAewAKACAAIAAgACAAcgBlAHQAdQByAG4AOwAKACAAIAB9AAoAIAAgAGYAbABhAGcAIAA9ACAAdAByAHUAZQA7AAoAfQAKAAoALwAvACAAQwBvAG4AdAByAG8AbAAgAEIAOgAgAGEAIABnAGUAbgB1AGkAbgBlAGwAeQAgAHUAbgBuAGUAYwBlAHMAcwBhAHIAeQAgAGMAbwBuAGQAaQB0AGkAbwBuAC4AIABCAG8AdABoACAAbABpAG4AdABlAHIAcwAgAHIAZQBwAG8AcgB0ACAAdABoAGkAcwAsAAoALwAvACAAdwBoAGkAYwBoACAAcwBoAG8AdwBzACAAdABoAGUAIAByAHUAbABlACAAaQBzACAAcwB3AGkAdABjAGgAZQBkACAAbwBuAC4ACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAGMAbwBuAHQAcgBvAGwAKABzADoAIABzAHQAcgBpAG4AZwApACAAewAKACAAIABpAGYAIAAoAHMAIAAhAD0APQAgAHUAbgBkAGUAZgBpAG4AZQBkACkAIAB7AAoAIAAgACAAIAByAGUAdAB1AHIAbgAgAHMAOwAKACAAIAB9AAoAIAAgAHIAZQB0AHUAcgBuACAAIgAiADsACgB9AAoA


https://typescript-eslint.io/play/#fileType=.tsx&code=PTAEGUBcFcCNQPYDNQCUCmBDAxpA5AM6jQHoZIA0oBCokAFgJZEB266AJq7QAKQCeAB3QFgAJyy4AdACgO6bABtME0EmgtcjBC2KlyAHgAqAPgAUjFo0iNMigGp3o6AFygjASjcBvUNmhiEiyQbkagAL4A3DIyIKAAjFKggsqWiLAAVgqQoIrW6GJ2oADu9AikyWIIwmICoMygALbQkJiQnLltBTLYOgQ5AObQKhygALygvv6B6MFuSHYVUTHoAB6CCLVqGlo6oGIaAPKa6GYekzKg9ShmQyNS00GQ596XV-voMGIs0Vfhb3cxBwHgEnuM6Ad0NF-rEwAAmJIMdDUTCNZGA0YAN1sejI6CQbiRaEkOQ4CDginQAFoCHBGtZkm12t8ZGsNlt1JobHsSOgAOLDIFnC5XXosfr1FgAMTyA3oOQmvPIZgWilIHl%2BHy%2BukwBH4mm2XO0ulpsHpkGFr3e11AFmlsvlIJmwReb2tEm1mr%2BbslMsYcsgTrBE0gkK9EP4IutoEwxUwDIAClV6aQpBIaIpMacNT7wmpLHZFJGrdHLH6A0HZgq1Isobm3ssYXEAMxJFjQNFiRjYD5IVnrTY5Tm7XS8gDC5OClreYolukV%2BnxZgADDmrh6Arph9zdLAO4Jp9bGDcWJWXVH3Z9N%2BH-tbT48q%2BD4prGzE4hPglVFKAAIKE%2BjIgQqLIooCADN2iA6qAsCWBwlgDPsbQAWIdD0JgUGCFUNQCLIlJDsoCETKqpDRP27JDjsO77EcJwAEKwfBh42iqBGupenoNm8SAEeCobONCb5gB%2BoYIN%2BtFuJgoADLM0CWOgRbECwbDYCIQFiJGYpwTuSS0QgDC5JYzJEBI5GocwFCwiUTDYPQ1BlMURBEgclL1EQBDFNYNkdDoshsoOhojn4OgiYoZgEG4-RdiwAxscxRAAIRjIqLDyEgckcLF65Xt81Avm8G45QARIVAlAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by shbernal** [comment]
The same root cause reaches class fields, which may be worth folding into this issue since your control case (a plain `let` binding) still behaves correctly. On 2.5.7, Windows, `suspicious/noUnnecessaryConditions` as the only enabled rule:

```ts
export class Once {
  #done = false;

  run(): void {
    if (this.#done) return; // "This condition is always falsy."
    this.#done = true;
  }
}

export class Workbook {
  fullCalcOnLoad = false;
}

export function calcPr(workbook: Workbook): string {
  return workbook.fullCalcOnLoad ? '<calcPr/>' : ''; // "This condition is always falsy."
}
```

```
c-class-field.ts:5:9 lint/suspicious/noUnnecessaryConditions ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × This condition is always falsy.

  > 5 │     if (this.#done) return;
      │         ^^^^^^^^^^

  i The value's type can never be truthy, so this check is redundant.

c-class-field.ts:15:10 lint/suspicious/noUnnecessaryConditions ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × This condition is always falsy.

  > 15 │   return workbook.fullCalcOnLoad ? '<calcPr/>' : '';
       │          ^^^^^^^^^^^^^^^^^^^^^^^

  i The value's type can never be truthy, so this check is redundant.

Found 2 errors.
```

Two things this adds to the object-property case above:

- **A private field is flagged even though the mutation is three lines below the read, in the same method.** `#done` cannot be assigned from outside the class at all, so every assignment to it is visible to the analyzer — the whole-program information needed to see `true` is right there.
- **A public field declared `= false` is flagged at every read**, including in another module, which is the ordinary "boolean option, off by default" shape. Callers set `workbook.fullCalcOnLoad = true` and the branch is live.

Both are the same shape as `guard.current`: the declared type is narrowed to the initializer's literal type and later assignments are not folded back in. Worth noting the diagnostic wording — "The value's type can never be truthy" — is a confident claim of dead code, so on a large codebase these read like real bug reports rather than style noise. This variant is 16 of the 148 diagnostics the rule produces on a ~500-file TypeScript project (class-based streaming parsers with `#committed` / `#scanned` / `#inRun` latch fields).

**Comment by harshmathurx** [comment]
I’ll take this. I’ll keep the change scoped to noUnnecessaryConditions and add focused coverage for the mutable property/ref-like cases before changing the analyzer.
